packages/server/src/users.ts

```diff
@@ -14,6 +14,25 @@ export interface PersistentUser extends UserResponse {
   username: string;
   password_hash: string;
   login_type: "persistent";
+  rating?: number;
+}
+
+export async function updateUserRating(id: string, new_rating: number): Promise<void> {
```

Comment:
Two thoughts:
1.) Currently this function doesn't allow the calling code to observe the result of this update function. I believe we can enable this by returning the Promise (see code example below).
2.) If `id` is the session id, then in my opinion the variable name should be more explicit, like `session_id`, otherwise I would expect that this is the user id.
On that note, why don't we pass the user id as input here? I think that might save us one data base query.

Example:
```
export async function updateUserRating(user_id: string, new_rating: number): Promise<UpdateResult<GuestUser | PersistentUser>> {
  return await usersCollection().updateOne(
    { _id: new ObjectId(user_id) },
    { $set: {rating: new_rating} }
  );
}

Then we can call this from outside and add error handling, e.g.:

`await updateUserRating(user_id, 1500).catch(error => console.error(error));`
```

Reply:
Because it's marked "async", I believe it is returning a promise (but of course this will be more clear when we have the calling code)

Reply:
In the example, if we use the UpdateResult Type directly from MongoDB then we should be able to just return `Promise<UpdateResult>` instead of `Promise<UpdateResult<GuestUser | PersistentUser>>` correct? 

Looking at the `updateOne` function, it doesn't return the updated user data, only the `UpdateResult` promise. So it may be misleading to specify that it is returning type GuestUser or PersistentUser.

Reply:
> Looking at the updateOne function, it doesn't return the updated user data, only the UpdateResult promise. So it may be misleading to specify that it is returning type GuestUser or PersistentUser.

Hm as far as I can see the return type is `Promise<UpdateResult<TSchema>>`. In our case `usersCollection()` returns `Collection<GuestUser | PersistentUser>` i.e. `TSchema = GuestUser | PersistentUser`.

Reply:
Ah, I got it. Either way, it won't affect the return type so we can include the TSchema part like you said.

Reply:
Not a huge deal, but I actually don't think we should return Mongo types like UpdateResult from functions in `users.ts`.  Basically I'd like to encapsulate such that if we ever tried to support a different DB we wouldn't need a bunch of API changes.  (also, I'm not even sure what to do with this result object as a caller)

Reply:
I think I get what you mean. Hypothetically, if we keep it as is and use the return value for something then we later decide to switch to a new db, the way we use the return value would have to be changed. Is this what you mean? 

There's really no real purpose of the result object other than just making sure that values were updated. But now that we know that the function works, we can remove the return value.

Reply:
It's ok to me if you remove the return value, I'm just throwing out thoughts because I believe it can be worth discussing things like this.

> Basically I'd like to encapsulate such that if we ever tried to support a different DB we wouldn't need a bunch of API changes.

Yeah it's a valid argument. My thought was / is that it's generally a good habit to have the most flexible / specific return type. You can always ignore the returned value.

> But now that we know that the function works, we can remove the return value.

Database operations / queries can have unexpected behaviour. It's possible that they fail unexpectedly, e.g. if something locks the entry you try to change. So in my opinion it's good to handle these things with care. It is a case where "it usually works", doesn't mean "it always works".

Reply:
> Hypothetically, if we keep it as is and use the return value for something then we later decide to switch to a new db, the way we use the return value would have to be changed. Is this what you mean?

Correct!

> There's really no real purpose of the result object other than just making sure that values were updated. But now that we know that the function works, we can remove the return value.

There may be a purpose, but if there is, we should massage the value into something more understandable if we want to return something from this function.

> Database operations / queries can have unexpected behaviour. It's possible that they fail unexpectedly, e.g. if something locks the entry you try to change. So in my opinion it's good to handle these things with care. It is a case where "it usually works", doesn't mean "it always works".

Absolutely!  @merowin are there any states you think we need to bubble up to the user? I don't have a good understanding of this return object myself (upsertId, upsertCount etc.), and my hope was that all failures would lead to a rejected `Promise`.

Reply:
> Absolutely! @merowin are there any states you think we need to bubble up to the user? I don't have a good understanding of this return object myself (upsertId, upsertCount etc.)

It seems that upsertId, upsertCount are only relevant when we pass the options { upsert: true } to the mutation function (e.g. updateOne, updateMany, maybe more). From [the docs](https://www.mongodbtutorial.org/mongodb-crud/mongodb-upsert/): 
```
Upsert is a combination of update and insert. 

Upsert performs two functions:

    Update data if there is a matching document.
    Insert a new document in case there is no document matches the query criteria.
```

I think the most interesting part about the UpdateResult might be `acknowledged`.

> all failures would lead to a rejected Promise.

This text in the node_modules files makes me think that we might just get the response `{ acknowledged: false }` in case of an error.

```
export declare interface UpdateResult<TSchema extends Document = Document> {
    /** Indicates whether this write result was acknowledged. If not, then all other members of this result will be undefined */
    acknowledged: boolean;
    [...]
}
```

Reply:
Thanks for helping me understand the upsert stuff!

> we might just get the response { acknowledged: false } in case of an error.

If it is an indication of failure, one approach may be to turn it into a rejected Promise:

```
const result = await updateOne(...);
if (!result.acknowledged) {
    // throw in an async function is the same as rejected Promise
    throw new Error("Update not acknowledged");
}
```

Though based on [these docs](https://www.mongodb.com/docs/manual/reference/method/db.collection.updateOne/#returns), it seems possible acknowledgement is just telling us whether or not write concern is enabled.

> A boolean acknowledged as true if the operation ran with [write concern](https://www.mongodb.com/docs/manual/reference/glossary/#std-term-write-concern) or false if write concern was disabled

Reply:
>Though based on [these docs](https://www.mongodb.com/docs/manual/reference/method/db.collection.updateOne/#returns), it seems possible acknowledgement is just telling us whether or not write concern is enabled.

After some testing, it seems that even if a faulty id is passed, `acknowledged`  is still true. I would say that a better indicator of failure would be `matchedCount` which tells us the number of documents that matched the filter. So we would have something like this:

```
if(result.matchedCount == 0){
    throw new Error("User not found")
  }
```